data-library/library/script/nycoc_checkbook.py

```diff
@@ -0,0 +1,448 @@
+    "Payroll",
+    "Spending",
+    "Spending_OGE",
+    "Contracts_OGE",
+    "Spending_NYCHA",
+    "Contracts_NYCHA",
+    "Payroll_NYCHA",
+]
+
+DEFAULT_RECORDS_FROM = "1"
+DEFAULT_MAX_RECORDS = "20000"
+
+DEFAULT_DATE_FORMAT = "%Y-%m-%d"
+DEFAULT_START_PERIOD = dt.date(2010, 1, 1).strftime(DEFAULT_DATE_FORMAT)
+DEFAULT_END_PERIOD = dt.date.today().strftime(DEFAULT_DATE_FORMAT)
+
+DEFAULT_DATA_PATH = Path(__file__).resolve().parent / "capital_spending_data"
+
+USER_AGENTS = [
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36",
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36",
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36",
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36",
+    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36",
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.1 Safari/605.1.15",
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 13_1) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.1 Safari/605.1.15",
+]
+user_agent = random.choice(USER_AGENTS)
+
+
+class CriteriaValue(TypedDict):
+    name: str
+    type: Literal["value"]
+    value: str
+
+
+class CriteriaRange(TypedDict):
+    name: str
+    type: Literal["range"]
+    start: str
+    end: str
+
+
+class Scriptor:
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+    def ingest(
+        self,
+        user_agent: str | None = None,
+        type_of_data: TYPE_OF_DATA = "Spending",
+        records_from: str = DEFAULT_RECORDS_FROM,
+        max_records: str = DEFAULT_MAX_RECORDS,
+        response_columns: list[str] | None = None,
+        search_criteria: list[CriteriaValue | CriteriaRange] | None = None,
+        num_retries: int = 3,
+        data_dir: Path = DEFAULT_DATA_PATH,
+        start_period: str = DEFAULT_START_PERIOD,
+        end_period: str = DEFAULT_END_PERIOD,
+    ) -> pd.DataFrame:
+        """Fetches data from Checkbook NYC API based on input parameters, saves it
+        locally in parquet files organized by months, and returns data in a pandas DataFrame.
+        """
+        create_data_dir(data_dir)
+
+        search_criteria = validate_search_criteria(search_criteria)
+
+        month_range = generate_monthly_ranges(start_period, end_period)
+
+        for month in month_range:
+            start, end = month
+            month_criteria: CriteriaRange = {
+                "name": "issue_date",
+                "type": "range",
+                "start": start,
+                "end": end,
+            }
+            search_criteria_modified = search_criteria + [month_criteria]
+            try:
+                df = get_data(
```

Comment:
`get_data` is the new function that gets ALL data for a given month (one iteration in this loop). If, for some reason, we can't get all records for a month period, an Exception is printed and the `for` loop continues to the next month.